resources/assets/lib/select-options.coffee

```diff
@@ -0,0 +1,119 @@
+#    attracting more community contributions to the core ecosystem of osu!.
+#
+#    osu!web is free software: you can redistribute it and/or modify
+#    it under the terms of the Affero GNU General Public License version 3
+#    as published by the Free Software Foundation.
+#
+#    osu!web is distributed WITHOUT ANY WARRANTY; without even the implied
+#    warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+#    See the GNU Affero General Public License for more details.
+#
+#    You should have received a copy of the GNU Affero General Public License
+#    along with osu!web.  If not, see <http://www.gnu.org/licenses/>.
+###
+
+import { a, i, div } from 'react-dom-factories'
+import { PureComponent } from 'react'
+
+export class SelectOptions extends PureComponent
+  constructor: (props) ->
+    super props
+    @bn = @props.bn ? 'select-options'
+
+    @state =
+      showingSelector: false
+
+
+  componentDidMount: =>
+    $(document).on "click.#{@bn}", @hideSelector
+
+
+  componentDidUpdate: (_prevProps, prevState) =>
+    Blackout.toggle(@state.showingSelector, 0.5) unless prevState.showingSelector == @state.showingSelector
+
+
+  componentWillUnmount: ->
+    $(document).off ".#{@bn}"
+
+
+  render: =>
+    classNames = "#{@bn} js-#{@bn}"
+    classNames += " #{@bn}--selecting" if @state.showingSelector
+
+    div
+      className: classNames
+      div
+        className: "#{@bn}__select"
+        @renderItem
+          children: [
+            div
+              className: 'u-ellipsis-overflow'
+              key: 'current'
+              @props.selected?.text
+
+            div
+              key: 'decoration'
+              className: "#{@bn}__decoration",
+              i className: "fas fa-chevron-down"
+          ]
+          item: @props.selected
+          onClick: @toggleSelector
+
+      div
+        className: "#{@bn}__selector"
+        for item in @props.options
+          @renderOption item
+
+
+  renderOption: (item) =>
+    @renderItem
+      children: [
+        div
+          className: 'u-ellipsis-overflow'
+          key: item.id
+          item.text
+      ],
+      item: item
+      selected: @props.selected?.id == item.id
+      onClick: (event) => @itemSelected(event, item)
+
+
+  renderItem: ({ children, item, onClick, selected = false }) ->
+    cssClasses = "#{@bn}__item"
+    cssClasses += " #{@bn}__item--selected" if selected
+
+    return @props.renderItem({ children, item, onClick, cssClasses }) if @props.renderItem?
+
+    a
+      children: children
+      className: cssClasses
+      href: '#'
+      key: item.id
+      onClick: onClick
+
+
+  hideSelector: (e) =>
+    return if e.button != 0
+    return unless @state.showingSelector
+    return if $(e.target).closest(".js-#{@bn}").length
+
+    @setState showingSelector: false
+
+
+  itemSelected: (event, item) ->
+    return if event.button != 0
+    event.preventDefault()
+
+    @setState showingSelector: false
```

Comment:
probably isn't needed if `.closest` check in `hideSelector` only covers the select box instead of whole thing including options.